chips/stm32f429zi/src/ethernet/ethernet.rs

```diff
@@ -0,0 +1,2288 @@
+        self.mac_rx.enable();
+        self.mac_tx.enable();
+        self.mac_ptp.enable();
+    }
+}
+
+/// Ethernet peripheral
+pub struct Ethernet<'a> {
+    mac_registers: StaticRef<Ethernet_MacRegisters>,
+    _mmc_registers: StaticRef<Ethernet_MmcRegisters>,
+    dma_registers: StaticRef<Ethernet_DmaRegisters>,
+    transmit_descriptor: TransmitDescriptor,
+    receive_descriptor: ReceiveDescriptor,
+    transmit_packet: TakeCell<'static, [u8]>,
+    transmit_packet_length: OptionalCell<u16>,
+    packet_identifier: OptionalCell<usize>,
+    received_packet: TakeCell<'a, [u8]>,
+    number_packets_missed: Cell<usize>,
+    client: OptionalCell<&'a dyn EthernetAdapterClient>,
+    clocks: EthernetClocks<'a>,
+    mac_address0: Cell<MacAddress>,
+}
+
+const DEFAULT_MAC_ADDRESS: MacAddress = MacAddress::new([0xD4, 0x5D, 0x64, 0x62, 0x95, 0x1A]);
+
+impl<'a> Ethernet<'a> {
+    /// Ethernet constructor
+    pub fn new(rcc: &'a rcc::Rcc) -> Self {
+        Self {
+            mac_registers: ETHERNET_MAC_BASE,
+            _mmc_registers: ETHERNET_MMC_BASE,
+            dma_registers: ETHERNET_DMA_BASE,
+            transmit_descriptor: TransmitDescriptor::new(),
+            receive_descriptor: ReceiveDescriptor::new(),
+            transmit_packet: TakeCell::empty(),
+            transmit_packet_length: OptionalCell::empty(),
+            packet_identifier: OptionalCell::empty(),
+            received_packet: TakeCell::empty(),
+            number_packets_missed: Cell::new(0),
+            client: OptionalCell::empty(),
+            clocks: EthernetClocks::new(rcc),
+            mac_address0: Cell::new(DEFAULT_MAC_ADDRESS),
+        }
+    }
+
+    fn init_transmit_descriptors(&self) {
+        self.transmit_descriptor.release();
+        self.transmit_descriptor.enable_interrupt_on_completion();
+        self.transmit_descriptor.set_as_first_segment();
+        self.transmit_descriptor.set_as_last_segment();
+        self.transmit_descriptor.enable_pad();
+        self.transmit_descriptor.enable_crc();
+        self.transmit_descriptor.set_transmit_end_of_ring();
+    }
+
+    fn init_receive_descriptors(&self) {
+        self.receive_descriptor.release();
+        self.receive_descriptor.enable_interrupt_on_completion();
+        self.receive_descriptor.set_receive_end_of_ring();
+    }
+
+    fn init_dma(&self) -> Result<(), ErrorCode> {
+        self.reset_dma()?;
+        self.flush_dma_transmit_fifo()?;
+        self.disable_flushing_received_frames();
+        self.forward_error_frames();
+        self.forward_undersized_good_frames();
+        self.enable_transmit_store_and_forward()?;
+        self.enable_receive_store_and_forward()?;
+
+        self.set_transmit_descriptor_list_address(
+            &self.transmit_descriptor as *const TransmitDescriptor as u32,
+        )?;
+        self.set_receive_descriptor_list_address(
+            &self.receive_descriptor as *const ReceiveDescriptor as u32,
+        )?;
```

Comment:
Note to self: This is safe, given that both `TransmitDescriptor` and `ReceiveDescriptor` are `#[repr(C)]` structs which only contain `InMemoryRegister`s, which in turn wrap `VolatileCell` -- so we don't need to main Rust's aliasing requirements here, and spurious reads are also not an issue.

Reply:
Added comment with your explanation.